packages/integrations/cloudflare/src/server.advanced.ts

```diff
@@ -47,17 +58,32 @@ export function createExports(manifest: SSRManifest) {
 				},
 			});
 
-			let response = await app.render(request, routeData, {
-				runtime: {
-					waitUntil: (promise: Promise<any>) => {
-						context.waitUntil(promise);
+			const cloudflareContext = {} as CloudflareContext;
+
+			// We define a custom property, so we can check the value passed to locals
+			Object.defineProperty(cloudflareContext, 'locals', {
+				enumerable: true,
+				// we should protect top-level locals from being overwritten
+				// users should just use nested properties, e.g. locals.test = "foo"
+				writable: false,
+				configurable: false,
+				value: {
+					runtime: {
+						waitUntil: (promise: Promise<any>) => {
+							context.waitUntil(promise);
+						},
+						env: env,
+						cf: request.cf,
+						caches: caches,
 					},
-					env: env,
-					cf: request.cf,
-					caches: caches,
 				},
 			});
 
+			// We should freeze the runtime object, to make sure it&nested properties are not mutated
+			Object.freeze(cloudflareContext.locals.runtime);
+
+			let response = await app.render(request, routeData, cloudflareContext.locals);
```

Comment:
this is a little different implemented than in the reference, but AFAIK this actually puts all the protections in place. Since using the adapter, locals now has a value all the time, the user should not ever write to `context.locals`!, even not in object.

Reply:
@ematipico let me know if that thought makes actually sense or not.